File: trunk/jshiny/src/jlo/ioe/util/LazyRef.java
package jlo.ioe.util;

import java.io.Serializable;
import java.util.concurrent.locks.Lock;
import java.util.concurrent.locks.ReentrantLock;

/**
 * Copyright © 2007 imaginaryday.com (jlo)<br>
 * User: jlowens<br>
 * Date: Apr 4, 2007<br>
 * Time: 6:50:05 PM<br>
 */
public abstract class LazyRef<T extends Identifiable> implements Serializable {
	transient private Opt<T> object = Opt.none();
	private ObjectID oid;
	transient private Lock lock = new ReentrantLock();

	public LazyRef(T object) {
		this.object = Opt.some(object);
		this.oid = object.getObjectID();
	}

	public LazyRef(ObjectID oid) {
		this.oid = oid;
		object = Opt.none();
	}

	// eagerly created, but lazily executed
	private F.lambda<Opt<T>> loader = new F.lambda0<Opt<T>>(){protected Opt<T> code() {
		return loadObject(oid);
	}};

	public T get() {
		lock.lock();
		try {
			return (object = object.orElse(loader)).get((T)null);
		} finally { lock.unlock(); }
	}

	public boolean isLoaded() {
		return isLoaded(oid);
	}

	public ObjectID getOID() {
		return oid;	
	}

	abstract protected boolean isLoaded(ObjectID oid);
	abstract protected Opt<T> loadObject(ObjectID oid);
}

File: trunk/sshiny/src/jlo/ioe/util/LazyRef.scala
package jlo.ioe.util

// todo: refactor to util
import jlo.ioe.data.ObjectID
import java.io.{ObjectInput,ObjectOutput}

@SerialVersionUID(1000)
abstract class LazyRef[T <: Identifiable](var obj : Option[T]) extends java.io.Externalizable {
  var oid : ObjectID = if (obj.isDefined) obj.get.objectID else null

  def this() = this(None)
  def this(id : ObjectID) = { this(None); oid = id }

  def objectID = oid
  def isLoaded : boolean = isLoaded(oid)
  def apply() : T = synchronized { obj = obj.orElse({loadObject(oid)}); obj.get }
  protected def isLoaded(id:ObjectID) : boolean
  protected def loadObject(id:ObjectID) : Option[T]
  
  def writeExternal(out:ObjectOutput) : Unit = {
    out.writeObject(oid)
  }
  def readExternal(in:ObjectInput) : Unit = {
    oid = in.readObject.asInstanceOf[ObjectID]
  }
}
